ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/worldstate/Pruner.java

```diff
@@ -0,0 +1,134 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.worldstate;
+
+import tech.pegasys.pantheon.ethereum.chain.BlockAddedEvent;
+import tech.pegasys.pantheon.ethereum.chain.Blockchain;
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.core.Hash;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class Pruner {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final MarkSweepPruner pruningStrategy;
+  private final Blockchain blockchain;
+  private final ExecutorServiceBuilder executorServiceBuilder;
+  private ExecutorService executor;
+  private final long retentionPeriodInBlocks;
+  private final AtomicReference<State> state = new AtomicReference<>(State.IDLE);
+  private volatile long markBlockNumber = 0;
+  private volatile BlockHeader markedBlockHeader;
+  private long transientForkThreshold;
+
+  public Pruner(
+      final MarkSweepPruner pruningStrategy,
+      final Blockchain blockchain,
+      final ExecutorServiceBuilder executorServiceBuilder,
+      final long transientForkThreshold,
+      final long retentionPeriodInBlocks) {
+    this.pruningStrategy = pruningStrategy;
+    this.executorServiceBuilder = executorServiceBuilder;
+    this.blockchain = blockchain;
+    if (!(transientForkThreshold >= 0 && retentionPeriodInBlocks >= transientForkThreshold)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Expected transientForkThreshold >= 0 and retentionPeriodInBlocks >= transientForkThreshold. transientForkThreshold=%d, retentionPeriodInBlocks=%d",
+              transientForkThreshold, retentionPeriodInBlocks));
+    }
+    this.retentionPeriodInBlocks = retentionPeriodInBlocks;
+    this.transientForkThreshold = transientForkThreshold;
+  }
+
+  public void start() {
+    executor = executorServiceBuilder.build();
+    blockchain.observeBlockAdded((event, blockchain) -> handleNewBlock(event));
+  }
+
+  public void stop() throws InterruptedException {
+    if (executor != null) {
+      executor.awaitTermination(30, TimeUnit.SECONDS);
+    }
+  }
+
+  private void handleNewBlock(final BlockAddedEvent event) {
+    if (!event.isNewCanonicalHead()) {
+      return;
+    }
+
+    final long blockNumber = event.getBlock().getHeader().getNumber();
+    if (state.compareAndSet(State.IDLE, State.TRANSIENT_FORK_OUTLIVING)) {
```

Comment:
In this stretch of if blocks is it ever possible to trigger two of them in the same call?  Would that be a bug?  If only one could fire should we `else if` them?  This could interact with my short-circuit boolean comment below.

Reply:
It used to be `else if`'d but I changed them to be chained in order to allow it to cascade from one state to another in, for example, the case where the `TRANSIENT_FORK_OUTLIVING` period is 0.

Reply:
That makes me wonder if the logic is too fragile.  Could we switch or if/else if on the current state and then check our other conditions and then set?

Reply:
To really make sure the logic is tight (we're exiting asap, not checking states we know we can't be in), we can use sometimes breakless cases in a switch to cascade through. I find that it has quite difficult to grasp the control flow though. The current version is doing unnecessary checks but we know that once we get one failure to state transition, the rest will fail too and it's much more readable.

```
    switch (state.get()) {
      case IDLE:
        {
          state.set(State.TRANSIENT_FORK_OUTLIVING);
          pruningStrategy.prepare();
          markBlockNumber = blockNumber;
        }
      case TRANSIENT_FORK_OUTLIVING:
        {
          if (blockNumber >= markBlockNumber + transientForkThreshold) {
            state.set(State.MARKING);
            markedBlockHeader = blockchain.getBlockHeader(markBlockNumber).get();
            mark(markedBlockHeader);
          } else break;
        }
      case MARKING_COMPLETE:
        {
          if (state.get() != State.MARKING_COMPLETE) // possible if we flow over from previous case
          break;
          if (blockNumber >= markBlockNumber + retentionPeriodInBlocks
              && blockchain.blockIsOnCanonicalChain(markedBlockHeader.getHash())) {
            state.set(State.SWEEPING);
            sweep();
          }
        }
    }
```

Reply:
Looks good.  
Style wise I would drop the braces for the case bodies.  Also I would include a `// nobreak` or `// fallthrough` line in the cases you are intending to fall through, for explicit readability.  And the `else break;` will require braces but spotless will catch it.

Reply:
After talking with Meredith, we decided that switching back to the `if/else if` and simply not allowing the case where you cascade through multiple states in one go because it probably won't matter to the user and it makes the code a lot cleaner.